src/Workspaces/Remote/Core/AbstractAssetProvider.cs

```diff
@@ -75,45 +97,100 @@ public async Task<ProjectInfo> CreateProjectInfoAsync(ProjectId projectId, Check
+    public async Task SynchronizeProjectDocumentsAsync(
+        ProjectStateChecksums projectChecksums, CancellationToken cancellationToken)
     {
-        var documentSnapshot = await GetAssetAsync<DocumentStateChecksums>(new(AssetPathKind.DocumentStateChecksums, documentId), documentChecksum, cancellationToken).ConfigureAwait(false);
-        Contract.ThrowIfTrue(documentId != documentSnapshot.DocumentId);
+        await Task.Yield();
+
+        using var _1 = PooledHashSet<Checksum>.GetInstance(out var attributeChecksums);
+        using var _2 = PooledHashSet<Checksum>.GetInstance(out var textChecksums);
+
+        projectChecksums.Documents.AttributeChecksums.AddAllTo(attributeChecksums);
+        projectChecksums.AdditionalDocuments.AttributeChecksums.AddAllTo(attributeChecksums);
+        projectChecksums.AnalyzerConfigDocuments.AttributeChecksums.AddAllTo(attributeChecksums);
 
-        var attributes = await GetAssetAsync<DocumentInfo.DocumentAttributes>(new(AssetPathKind.DocumentAttributes, documentId), documentSnapshot.Info, cancellationToken).ConfigureAwait(false);
-        var serializableSourceText = await GetAssetAsync<SerializableSourceText>(new(AssetPathKind.DocumentText, documentId), documentSnapshot.Text, cancellationToken).ConfigureAwait(false);
+        projectChecksums.Documents.TextChecksums.AddAllTo(textChecksums);
+        projectChecksums.AdditionalDocuments.TextChecksums.AddAllTo(textChecksums);
+        projectChecksums.AnalyzerConfigDocuments.TextChecksums.AddAllTo(textChecksums);
+
+        var attributesTask = this.GetAssetsAsync<DocumentInfo.DocumentAttributes>(
+            assetPath: new(AssetPathKind.DocumentAttributes, projectChecksums.ProjectId),
+            attributeChecksums,
+            cancellationToken);
+
+        var textTask = this.GetAssetsAsync<SerializableSourceText>(
+            assetPath: new(AssetPathKind.DocumentText, projectChecksums.ProjectId),
+            textChecksums,
+            cancellationToken);
+
+        await Task.WhenAll(attributesTask, textTask).ConfigureAwait(false);
+    }
+
+    public async Task<DocumentInfo> CreateDocumentInfoAsync(
+        DocumentId documentId, Checksum attributeChecksum, Checksum textChecksum, CancellationToken cancellationToken)
+    {
+        var attributes = await GetAssetAsync<DocumentInfo.DocumentAttributes>(new(AssetPathKind.DocumentAttributes, documentId), attributeChecksum, cancellationToken).ConfigureAwait(false);
+        var serializableSourceText = await GetAssetAsync<SerializableSourceText>(new(AssetPathKind.DocumentText, documentId), textChecksum, cancellationToken).ConfigureAwait(false);
 
         var text = await serializableSourceText.GetTextAsync(cancellationToken).ConfigureAwait(false);
         var textLoader = TextLoader.From(TextAndVersion.Create(text, VersionStamp.Create(), attributes.FilePath));
 
         // TODO: do we need version?
```

Comment:
Out of curiosity, is this still an unknown?

Reply:
can discuss offline.